src/server/ua_server.c

```diff
@@ -616,6 +616,120 @@ UA_Server_createServerConnection(UA_Server *server, const UA_String *serverUrl)
+                                 UA_Server_reverseConnectCallback);
+
+        UA_KeyValueMap_clear(&params);
+
+        if (res != UA_STATUSCODE_GOOD) {
+            context->currentConnection.connectionId = 0;
+        }
+
+        if (context->state != UA_SECURECHANNELSTATE_CONNECTING) {
+            context->state = UA_SECURECHANNELSTATE_CONNECTING;
+            if (context->stateCallback)
+                context->stateCallback(server, context->handle, context->state,
+                                       context->callbackContext);
+        }
+    }
+
+    return res;
+}
+
+UA_StatusCode UA_Server_addReverseConnect(UA_Server *server, UA_String url,
+                                          UA_Server_ReverseConnectStateCallback stateCallback,
+                                          void *callbackContext, UA_UInt64 *handle) {
+    UA_ServerConfig *config = UA_Server_getConfig(server);
+
+    UA_String hostname = UA_STRING_NULL;
+    UA_UInt16 port = 0;
+    UA_StatusCode res = UA_parseEndpointUrl(&url, &hostname, &port, NULL);
+    if(res != UA_STATUSCODE_GOOD) {
+        UA_LOG_WARNING(&config->logger, UA_LOGCATEGORY_SERVER,
+                       "OPC UA URL is invalid: %.*s",
+                       (int)url.length, url.data);
+        return res;
+    }
+
+    if (SLIST_EMPTY(&server->reverseConnects))
+        UA_Server_setReverseConnectRetryCallback(server, true);
+
+    reverse_connect_context *newContext = (reverse_connect_context *)calloc(1, sizeof(reverse_connect_context));
+    UA_String_copy(&hostname, &newContext->hostname);
+    newContext->port = port;
+    newContext->handle = ++server->lastReverseConnectHandle;
+    newContext->stateCallback = stateCallback;
+    newContext->callbackContext = callbackContext;
+    SLIST_INSERT_HEAD(&server->reverseConnects, newContext, next);
+
+    if (handle)
+        *handle = newContext->handle;
+
+    return UA_Server_attemptReverseConnect(server, newContext);
+}
+
+UA_StatusCode UA_Server_removeReverseConnect(UA_Server *server, UA_UInt64 handle) {
+    reverse_connect_context *rev = NULL;
+    reverse_connect_context *temp = NULL;
+    UA_StatusCode result = UA_STATUSCODE_BADNOTFOUND;
+
+    SLIST_FOREACH_SAFE(rev, &server->reverseConnects, next, temp) {
+        if (rev->handle == handle) {
+            SLIST_REMOVE(&server->reverseConnects, rev, reverse_connect_context, next);
+
+            if (rev->currentConnection.connectionId) {
+                rev->destruction = true;
+                /* Request disconnect and run the event loop for one iteration */
+                if (rev->currentConnection.connectionId) {
+                    rev->currentConnection.connectionManager->closeConnection(
+                                rev->currentConnection.connectionManager,
+                                rev->currentConnection.connectionId);
+                    server->config.eventLoop->run(server->config.eventLoop, 0);
```

Comment:
@jpfr  IMO the user should also be able to kick out a connected client by removing the associated reverse connect.
Would it be possible to add a deferred callback to delete the context or would that be run in the next event loop iteration before the socket has been closed?

Reply:
Yes!
What can be done is to remove the reverse connect from the linked list.
And add a "delayed callback" to the eventloop.
That will be automatically executed in the next iteration of the eventloop.
So that nobody else uses the pointer.